server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,226 @@
+    gnome::secret_exchange::SecretExchange,
+    prompt::{Prompt, PromptRole},
+    service::Service,
+};
+
+// System prompt properties
+#[derive(DeserializeDict, SerializeDict, Type)]
+#[zvariant(signature = "dict")]
+pub struct Properties {
+    title: String,
+    #[zvariant(rename = "choice-label")]
+    choice_label: String,
+    description: String,
+    message: String,
+    #[zvariant(rename = "caller-window")]
+    caller_window: String,
+    warning: String,
+    #[zvariant(rename = "password-new")]
+    password_new: bool,
+    #[zvariant(rename = "choice-chosen")]
+    choice_chosen: bool,
+    #[zvariant(rename = "continue-label")]
+    continue_label: String,
+    #[zvariant(rename = "cancel-label")]
+    cancel_label: String,
+}
+
+// org.gnome.keyring.internal.Prompter
+
+#[proxy(
+    default_service = "org.gnome.keyring.SystemPrompter",
+    interface = "org.gnome.keyring.internal.Prompter",
+    default_path = "/org/gnome/keyring/Prompter"
+)]
+pub trait Prompter {
+    fn begin_prompting(&self, callback: &OwnedObjectPath) -> Result<(), ServiceError>;
+
+    fn perform_prompt(
+        &self,
+        callback: OwnedObjectPath,
+        type_: &str,
+        properties: Properties,
+        exchange: &str,
+    ) -> Result<(), ServiceError>;
+
+    fn stop_prompting(&self, callback: OwnedObjectPath) -> Result<(), ServiceError>;
+}
+
+// org.gnome.keyring.internal.Prompter.Callback
+
+pub struct PrompterCallback {
+    service: Service,
+    path: OwnedObjectPath,
+}
+
+#[interface(name = "org.gnome.keyring.internal.Prompter.Callback")]
+impl PrompterCallback {
+    pub async fn prompt_ready(
+        &self,
+        reply: &str,
+        properties: HashMap<&str, OwnedValue>,
+        exchange: &str,
+        #[zbus(connection)] connection: &zbus::Connection,
+    ) -> Result<(), ServiceError> {
+        let Some(prompt) = self.service.prompt().await else {
+            return Err(ServiceError::NoSuchObject(
+                "Prompt does not exist.".to_string(),
+            ));
+        };
+
+        match prompt.role() {
+            PromptRole::Lock => {
+                if properties.is_empty() {
+                    // First PromptReady call
+                    let secret_exchange = SecretExchange::new().map_err(|err| {
+                        ServiceError::ZBus(zbus::Error::FDO(Box::new(zbus::fdo::Error::Failed(
+                            format!("Failed to generate SecretExchange {err}."),
+                        ))))
+                    })?;
+                    let exchange = secret_exchange.begin();
+
+                    let properties = Properties {
+                        title: "".to_string(),
+                        choice_label: "".to_string(),
+                        description: "Confirm locking \"login\", Keyring.".to_string(),
+                        message: "Lock Keyring".to_string(),
+                        caller_window: "".to_string(),
+                        warning: "".to_string(),
```

Comment:
Optional ones should be `Option<String>`.

Reply:
Done, thanks!